docs/proposals/api-group.md

```diff
@@ -0,0 +1,104 @@
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Supporting multiple API groups
+
+## Goal
+
+1. Breaking the monolithic v1 API to modular groups
+
+2. Supporting different versions in different groups
+
+3. Supporting identically named kinds to exist in different groups
+
+4. Enabling/disabling groups individually
+
+5. Laying the basis for [API Plugin](../../docs/design/extending-api.md)
+
+6. Allowing users to omit group name when using kubectl
+
+
+## Bookkeeping for groups
+
+1. No changes to API spec:
+
+  We will not add new field in the API spec to hold the group information. Group will be part of the TypeMeta.APIVersion field, in the form of `group/version`.
+
+  For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid.
+
+2. conversion.Scheme:
+
+  The key of conversion.Scheme.versionMap will be "group/version". The internal version of objects of all groups will be registered to versionMap[""].
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified, the function should search groups in a predetermined order.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+    * for api/v1, we set the prefix="api" (which is populated from app.APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1
+
+    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
+
+    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
+
+2. handling common API objects like api.Status:
+
+  To handle the API objects that are used by all groups, we either have to register them to all schemes, or we can choose not to encode them to a version.
+
+  Currently, for api.Status, we choose the latter approach: the server will encode it if the codec is v1, otherwise the server will send it unversioned.
+
+## Client-side implementation
+
+1. Client struct:
+
+  Currently each group has a client struct wrapping over \*RESTClient, e.g., client/unversioned.ExperimentalClient, client/unversioned.Client. The reason we have separate client struct is because each of them also implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). We don't know yet how to make their implementation scalable.
```

Comment:
We don't really need those shims, that's why resource.Helper exists.  It
already implements those.

On Thu, Aug 20, 2015 at 12:55 PM, Muhammed Uluyol <notifications@github.com>
wrote:

> In docs/proposals/api-group.md
> <https://github.com/kubernetes/kubernetes/pull/12951#discussion_r37553239>
> :
>
> > +
> > +    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
> > +
> > +    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
> > +
> > +2. handling common API objects like api.Status:
> > +
> > +  To handle the API objects that are used by all groups, we either have to register them to all schemes, or we can choose not to encode them to a version.
> > +
> > +  Currently, for api.Status, we choose the latter approach: the server will encode it if the codec is v1, otherwise the server will send it unversioned.
> > +
> > +## Client-side implementation
> > +
> > +1. Client struct:
> > +
> > +  Currently each group has a client struct wrapping over \*RESTClient, e.g., client/unversioned.ExperimentalClient, client/unversioned.Client. The reason we have separate client struct is because each of them also implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). We don't know yet how to make their implementation scalable.
>
> An approach that may work is registering client api sets (I say sets
> because of plugins) with the client package.
>
> I imagine that we'd have a struct that needs to be populated for every
> type:
>
> type ClientAPI struct { // I don't like this name
>     // I'm assuming group, version, prefix, etc info is in the name
>     // if we need it. We could always have multiple fields or add a
>     // few parameters when we register a set of ClientAPI structs.
>     name   string
>     List   List(selector labels.Selector, list interface{}) error
>     Get    Get(name string, obj interface{}) error
>     Delete Delete(name string) error
>     Create Create(in, out interface{}) error
>     Update Update(in, out interface{}) error
>     Watch  Watch(label labels.Selector, field fields.Selector, resourceVersion string) (watch.Interface, error)
> }
>
> The on a client we'd have something like
>
> import (
>     kclient "k8s.io/kubernetes/client"
>     _ "k8s.io/kubernetes/client/v1"
>     _ "k8s.io/kubernetes/client/experimental"
>
>     osclient "github.com/openshift/origin/client"
> )
> [...]
>     registered, err := client.Discover(osclient.PluginSets)
> [...]
>     var podList api.PodList
>     err := client.ForNamespace(ns).List("pod", selector, &podList)
>
> The (un)boxing from/to interface{} isn't really noticed from the client's
> perspective which is nice, and since it's all registered, plugin additions
> are easy to support.
>
> However, much of the current client consists of duplicate shims, perhaps
> those could be combined which would make the registration unnecessary. Then
> there's the fact that PodInterface also has Bind(), which I haven't put any
> thought into.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12951/files#r37553239>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift